### PR TITLE
Update my fannkuch study version to be based on release version

### PR DESCRIPTION
**Pull request by bradcray**
As a means of reviewing the release version of fannkuch, I made stylistic and coding
changes to it and replaced the study version in my directory with it since we want to
focus on parallel versions over serial ones and I've got other serial variations that
are similar enough to my main study version.

Main changes from the release version in approximate order of importance:

* explicitly serialized whole-array operations on permutation vectors to avoid additional parallelism
* changed fannkuch() to use a for loop rather than a while loop
* made fannkuch take a range as its argument and simplified some off-by-one math
* leaned more heavily on type inference
* removed spaces before colons
* reformatted to fit 80 columns
* combined most adjacent declarations into a single one
* reordered declarations within main() to put them closer to their uses
* computed fact() outside of main() as part of module initialization.  Changed uses of fact() to use parens rather than square brackets in order to suggest that we could be computing it on the fly.  Also tried to initialize fact() with a serial iterator but it generates a warning :(  Also tried to move fact to the end of the file, but that generates an error (perhaps incorrectly?).  I remain unhappy with much of this...

I also started a list of possible TODOs in the README.

As part of this effort, I also had to change my existing study versions to use 'n' rather than 'N' so they could continue to share a PERFEXECOPTS file.

### CODE REVIEW

**Comment by bradcray** [comment]
@chapel-lang/perf-team : For your consideration.  The performance improvement is probably due primarily to rewriting implicitly parallel short-array operations as explicit serial for loops:  http://chapel.sourceforge.net/perf/shootout/?startdate=2016/06/01&enddate=2016/06/29&graphs=fannkuchreduxn12

File: test/release/examples/benchmarks/shootout/README [hunk 1]
```diff
@@ -106,6 +106,30 @@ o Chapel has discussed supporting some standard record-wrapped class
...
+
+o make scans not generate warnings and use them to compute/initialize
+  fact?
+
+o remove warning for serial whole-array assignment when RHS is serial
+  and then use an initfact() iterator / procedure to initialize fact[]
+  at declaration time
+
+o should globals be allowed to be declared below main()?  If so, move
+  fact[] to the bottom of the file?
```
Comment (lydia-duncan) [inline test/release/examples/benchmarks/shootout/README]: Whoa, we don't allow that?  That seems silly

File: test/studies/shootout/fannkuch-redux/bradc/fannkuch-redux-blc.chpl [hunk 2]
```diff
@@ -1,78 +1,119 @@
...
+config const n = 7,          // the array size over which to compute perms
+             nchunks = 720;  // the number of chunks of parallelism
+
+//
+// memoize n! (n factorial)
+//
+var fact: [0..n] int;
+fact[0] = 1;
+for i in 1..n do
+  fact[i] = i*fact[i-1];
```
Comment (lydia-duncan) [inline test/studies/shootout/fannkuch-redux/bradc/fannkuch-redux-blc.chpl]: Why do we want to initialize fact in the module initializer as opposed to main?  This strikes me as potentially confusing to new users trying to understand the order of execution for bits of the file (even with this initialization occurring before the main function is encountered).

File: test/studies/shootout/fannkuch-redux/bradc/fannkuch-redux-blc.chpl [hunk 3]
```diff
@@ -1,78 +1,119 @@
...
-  writeln(checksum);
-  writeln("Pfannkuchen(", N, ") = ", maxFlips);
+  var checkSum = 0,
+      maxFlips = 1;
+
+  var chunksz = (fact(n) + nchunks - 1) / nchunks;
+  chunksz += chunksz%2;
+  const work = 0..(fact(n) - chunksz) by chunksz;
+
+  forall lo in dynamic(work, 1) with (+ reduce checkSum,
```
Comment (lydia-duncan) [inline test/studies/shootout/fannkuch-redux/bradc/fannkuch-redux-blc.chpl]: lo represents low?

File: test/studies/shootout/fannkuch-redux/bradc/fannkuch-redux-blc.chpl [hunk 4]
```diff
@@ -1,78 +1,119 @@
...
-  return (checksum, maxFlips);
-}
 
+  proc firstPerm() {
+    var idx = inds.low;
+    for i in 1..n-1 by -1 {
+      const fact_i = fact(i),
+            d = idx / fact_i;
+      count[i] = d;
+      idx %= fact(i);
```
Comment (lydia-duncan) [inline test/studies/shootout/fannkuch-redux/bradc/fannkuch-redux-blc.chpl]: You just stored fact(i) in the constant fact_i, did you mean to reuse that variable here?

**Comment by lydia-duncan** [comment]
The explicitly serialized array assignments make me sad, but I suspect we don't want to submit this version?

File: test/studies/shootout/fannkuch-redux/bradc/fannkuch-redux-blc.chpl (hunk 2 above)
Comment (bradcray) [inline test/studies/shootout/fannkuch-redux/bradc/fannkuch-redux-blc.chpl]: I was pretty unhappy with this, but couldn't find a warning-free way to do it that night.  I figured out a way to make it work in a separate PR ( #4101) which I think you'll like better.

(I wanted it out of main() because it seemed more about setting up fact[] than it did about the "main" computation).

File: test/studies/shootout/fannkuch-redux/bradc/fannkuch-redux-blc.chpl (hunk 4 above)
Comment (bradcray) [inline test/studies/shootout/fannkuch-redux/bradc/fannkuch-redux-blc.chpl]: Probably.  Though I'll also look into the perf impact of caching it vs. indexing it twice.

**Comment by bradcray** [comment]
Ben suggested that one of the two serialized assignments could be changed back without impact (the slice one, I think?).  I was going to verify that.  But generally, yeah, I was going to propose submitting something like this version in spite of the manually serialized assignments.
